x/incentive/keeper/reward_tracker_events.go

```diff
@@ -0,0 +1,156 @@
+func (k Keeper) AddEventBtcDelegationUnbonded(ctx context.Context, height uint64, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	fpAddr, delAddr := fp.String(), del.String()
+
+	newEv := types.NewEventBtcDelegationUnboned(fpAddr, delAddr, amtSat)
+	k.Logger(ctx).Debug(
+		"add reward tracker events - unbonded",
+		"blockHeight", height,
+		"fpAddr", fp.String(),
+		"btcDelAddr", del.String(),
+	)
+
+	return k.AddRewardTrackerEvent(ctx, height, newEv)
+}
+
+// ProcessRewardTrackerEvents process all the reward tracker events from the latest processed height + 1
+// until the given block height.
+func (k Keeper) ProcessRewardTrackerEvents(ctx context.Context, untilBlkHeight uint64) error {
+	lastProcessedHeight, err := k.GetRewardTrackerEventLastProcessedHeight(ctx)
+	if err != nil {
+		return err
+	}
+
+	for blkHeight := lastProcessedHeight + 1; blkHeight <= untilBlkHeight; blkHeight++ {
+		if err := k.ProcessRewardTrackerEventsAtHeight(ctx, blkHeight); err != nil {
+			return err
+		}
+	}
+
+	return k.SetRewardTrackerEventLastProcessedHeight(ctx, untilBlkHeight)
+}
+
+// ProcessRewardTrackerEventsAtHeight gets all the events for that block height, process those events updating
+// the reward tracker structures and deletes all the events processed.
+// Note: if there is no event at that block height it returns nil.
+func (k Keeper) ProcessRewardTrackerEventsAtHeight(ctx context.Context, height uint64) error {
+	evts, err := k.GetOrNewRewardTrackerEvent(ctx, height)
+	if err != nil {
+		return err
+	}
+
+	k.Logger(ctx).Debug(
+		"processing reward tracker events",
+		"blockHeight", height,
+		"events", len(evts.Events),
+	)
+
+	// it must process all the events without pagination
+	for _, untypedEvt := range evts.Events {
+		switch typedEvt := untypedEvt.Ev.(type) {
+		case *types.EventPowerUpdate_BtcActivated:
+			evt := typedEvt.BtcActivated
+			fp, del := sdk.MustAccAddressFromBech32(evt.FpAddr), sdk.MustAccAddressFromBech32(evt.BtcDelAddr)
+			if err := k.BtcDelegationActivated(ctx, fp, del, evt.TotalSat); err != nil {
+				return err
+			}
+		case *types.EventPowerUpdate_BtcUnbonded:
+			evt := typedEvt.BtcUnbonded
+			fp, del := sdk.MustAccAddressFromBech32(evt.FpAddr), sdk.MustAccAddressFromBech32(evt.BtcDelAddr)
+			if err := k.BtcDelegationUnbonded(ctx, fp, del, evt.TotalSat); err != nil {
+				return err
+			}
+		}
+	}
+
+	return k.DeleteRewardTrackerEvents(ctx, height)
+}
+
+// GetOrNewRewardTrackerEvent returns a new reward tracker if it doesn't exists for that block height
+func (k Keeper) GetOrNewRewardTrackerEvent(ctx context.Context, height uint64) (*types.EventsPowerUpdateAtHeight, error) {
+	found, err := k.rewardTrackerEvents.Has(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+
+	if !found {
+		return &types.EventsPowerUpdateAtHeight{
+			Events: make([]*types.EventPowerUpdate, 0),
```

Comment:
I guess this means that just after upgrade:
- few first blocks (the ones without events) will be rewarded against latest satte
- after some time we will switch to relaying on the events

Reply:
correct